src/graph/immutable_graph.cc

```diff
@@ -80,51 +110,70 @@ CSR::CSR(IdArray indptr, IdArray indices, IdArray edge_ids)
-  //   the sm_array because the refcount is maintained by the view arrays.
-  indptr_ = sm_array.CreateView({num_verts + 1}, DLDataType{kDLInt, 64, 1});
-  indices_ = sm_array.CreateView({num_edges}, DLDataType{kDLInt, 64, 1},
-      (num_verts + 1) * sizeof(dgl_id_t));
-  edge_ids_ = sm_array.CreateView({num_edges}, DLDataType{kDLInt, 64, 1},
-      (num_verts + 1 + num_edges) * sizeof(dgl_id_t));
+CSR::CSR(IdArray indptr, IdArray indices, IdArray edge_ids, bool is_multigraph,
+         const std::string &shared_mem_name): is_multigraph_(is_multigraph) {
+  CHECK(IsValidIdArray(indptr));
+  CHECK(IsValidIdArray(indices));
+  CHECK(IsValidIdArray(edge_ids));
+  CHECK_EQ(indices->shape[0], edge_ids->shape[0]);
+  const int64_t num_verts = indptr->shape[0] - 1;
+  const int64_t num_edges = indices->shape[0];
+  std::tie(indptr_, indices_, edge_ids_) = MapFromSharedMemory(
+      shared_mem_name, num_verts, num_edges);
   // copy the given data into the shared memory arrays
   indptr_.CopyFrom(indptr);
   indices_.CopyFrom(indices);
   edge_ids_.CopyFrom(edge_ids);
-#else
-  LOG(FATAL) << "CSR graph doesn't support shared memory in Windows yet";
-#endif  // _WIN32
 }
 
 CSR::CSR(const std::string &shared_mem_name,
-         int64_t num_verts, int64_t num_edges) {
-#ifndef _WIN32
-  const int64_t file_size = (num_verts + 1 + num_edges * 2) * sizeof(dgl_id_t);
-  IdArray sm_array = IdArray::EmptyShared(
-      shared_mem_name, {file_size}, DLDataType{kDLInt, 8, 1}, DLContext{kDLCPU, 0}, true);
-  // Create views from the shared memory array. Note that we don't need to save
-  //   the sm_array because the refcount is maintained by the view arrays.
-  indptr_ = sm_array.CreateView({num_verts + 1}, DLDataType{kDLInt, 64, 1});
-  indices_ = sm_array.CreateView({num_edges}, DLDataType{kDLInt, 64, 1},
-      (num_verts + 1) * sizeof(dgl_id_t));
-  edge_ids_ = sm_array.CreateView({num_edges}, DLDataType{kDLInt, 64, 1},
-      (num_verts + 1 + num_edges) * sizeof(dgl_id_t));
-#else
-  LOG(FATAL) << "CSR graph doesn't support shared memory in Windows yet";
-#endif  // _WIN32
+         int64_t num_verts, int64_t num_edges, bool is_multigraph)
+  : is_multigraph_(is_multigraph) {
+  std::tie(indptr_, indices_, edge_ids_) = MapFromSharedMemory(
+      shared_mem_name, num_verts, num_edges);
+}
+
+bool CSR::IsMultigraph() const {
+  return const_cast<CSR*>(this)->is_multigraph_.Get([this] () {
+      const dgl_id_t* indptr_data = static_cast<dgl_id_t*>(indptr_->data);
+      const dgl_id_t* indices_data = static_cast<dgl_id_t*>(indices_->data);
+      for (dgl_id_t src = 0; src < NumVertices(); ++src) {
+        std::unordered_set<dgl_id_t> hashmap;
+        for (dgl_id_t eid = indptr_data[src]; eid < indptr_data[src+1]; ++eid) {
+          const dgl_id_t dst = indices_data[eid];
+          if (hashmap.count(dst)) {
+            return true;
+          } else {
+            hashmap.insert(dst);
+          }
+        }
+      }
+      return false;
+    });
```

Comment:
please add comments here. `is_multigraph_` is initialized with a value. Is the lambda always called regardless of the initial value of `is_multigraph_`?

Reply:
done